wgpu-hal/src/dx12/shader_compilation.rs

```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
+        _dxc: dxc,
+        _dxil: dxil,
+    }))
+}
 
-        (result, log_level)
-    }
+/// Owned PCWSTR
+#[allow(clippy::upper_case_acronyms)]
+struct OPCWSTR {
+    inner: Vec<u16>,
+}
 
-    impl From<hassle_rs::HassleError> for crate::DeviceError {
-        fn from(value: hassle_rs::HassleError) -> Self {
-            match value {
-                hassle_rs::HassleError::Win32Error(e) => {
-                    // TODO: This returns an HRESULT, should we try and use the associated Windows error message?
-                    log::error!("Win32 error: {e:?}");
-                    crate::DeviceError::Lost
-                }
-                hassle_rs::HassleError::LoadLibraryError { filename, inner } => {
-                    log::error!("Failed to load dxc library {filename:?}. Inner error: {inner:?}");
-                    crate::DeviceError::Lost
-                }
-                hassle_rs::HassleError::LibLoadingError(e) => {
-                    log::error!("Failed to load dxc library. {e:?}");
-                    crate::DeviceError::Lost
-                }
-                hassle_rs::HassleError::WindowsOnly(e) => {
-                    log::error!("Signing with dxil.dll is only supported on Windows. {e:?}");
-                    crate::DeviceError::Lost
-                }
-                // `ValidationError` and `CompileError` should never happen in a context involving `DeviceError`
-                hassle_rs::HassleError::ValidationError(_e) => unimplemented!(),
-                hassle_rs::HassleError::CompileError(_e) => unimplemented!(),
-            }
-        }
+impl OPCWSTR {
+    fn new(s: &str) -> Self {
+        let mut inner: Vec<_> = s.encode_utf16().collect();
+        inner.push(0);
+        Self { inner }
     }
 
-    fn get_error_string_from_dxc_result(
-        library: &hassle_rs::DxcLibrary,
-        error: &hassle_rs::DxcOperationResult,
-    ) -> Result<String, hassle_rs::HassleError> {
-        error
-            .get_error_buffer()
-            .and_then(|error| library.get_blob_as_string(&hassle_rs::DxcBlob::from(error)))
+    fn ptr(&self) -> PCWSTR {
+        PCWSTR(self.inner.as_ptr())
     }
 }
 
-// These are stubs for when the `dxc_shader_compiler` feature is disabled.
-#[cfg(not(feature = "dxc_shader_compiler"))]
-mod dxc {
-    use std::ffi::CStr;
-    use std::path::PathBuf;
-
-    pub(crate) struct DxcContainer {}
-
-    pub(crate) fn get_dxc_container(
-        _dxc_path: Option<PathBuf>,
-        _dxil_path: Option<PathBuf>,
-    ) -> Result<Option<DxcContainer>, crate::DeviceError> {
-        // Falls back to Fxc and logs an error.
-        log::error!("DXC shader compiler was requested on Instance creation, but the DXC feature is disabled. Enable the `dxc_shader_compiler` feature on wgpu_hal to use DXC.");
-        Ok(None)
+fn get_output<T: Interface>(
+    res: &Dxc::IDxcResult,
+    kind: Dxc::DXC_OUT_KIND,
+) -> Result<T, crate::DeviceError> {
+    let mut result__: Option<T> = None;
```

Comment:
May just be me who prefers this short-hand:

```suggestion
    let mut result__ = None::<T>;
```

Although the type can be inferred:

```suggestion
    let mut result__ = None;
```

Not that it matters too much, this signature will change when upstream metadata fixes land.